File: android/src/main/kotlin/cl/ceisufro/native_video_view/NativeVideoViewController.kt
package cl.ceisufro.native_video_view

import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Build
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.VideoView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import java.util.*


class NativeVideoViewController(
    private val id: Int,
    private val context: Context,
    binaryMessenger: BinaryMessenger,
    private val lifecycleProvider: LifecycleProvider
) : DefaultLifecycleObserver,
    MethodChannel.MethodCallHandler,
    PlatformView,
    MediaPlayer.OnPreparedListener,
    MediaPlayer.OnErrorListener,
    MediaPlayer.OnCompletionListener {
    private val methodChannel: MethodChannel =
        MethodChannel(binaryMessenger, "native_video_view_$id")
    private val lifeCycleHashcode: Int
    private val constraintLayout: ConstraintLayout
    private var videoView: VideoView? = null
    private var dataSource: String? = null
    private var disposed: Boolean = false
    private var requestAudioFocus: Boolean = true
    private var volume: Double = 1.0
    private var mute: Boolean = false
    private var mediaPlayer: MediaPlayer? = null
    private var playerState: PlayerState = PlayerState.NOT_INITIALIZED

    init {
        this.methodChannel.setMethodCallHandler(this)
        this.lifeCycleHashcode = lifecycleProvider.getLifecycle().hashCode()
        this.constraintLayout = LayoutInflater.from(context)
            .inflate(R.layout.video_layout, null) as ConstraintLayout
        lifecycleProvider.getLifecycle()!!.addObserver(this)
    }

    override fun getView(): View {
        return constraintLayout
    }

    override fun dispose() {
        if (disposed) return
        disposed = true
        methodChannel.setMethodCallHandler(null)
        this.destroyVideoView()
        lifecycleProvider.getLifecycle()!!.removeObserver(this)
        Log.d("NVV#NativeView", "Disposed view $id")
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "player#setVideoSource" -> {
                val videoPath: String? = call.argument("videoSource")
                val sourceType: String? = call.argument("sourceType")
                requestAudioFocus = call.argument("requestAudioFocus") as Boolean? ?: true
                if (videoPath != null) {
                    if (sourceType.equals("VideoSourceType.asset")
                        || sourceType.equals("VideoSourceType.file")
                    ) {
                        initVideo("file://$videoPath")
                    } else {
                        initVideo(videoPath)
                    }
                }
                result.success(null)
            }
            "player#start" -> {
                startPlayback()
                result.success(null)
            }
            "player#pause" -> {
                pausePlayback()
                result.success(null)
            }
            "player#stop" -> {
                stopPlayback()
                result.success(null)
            }
            "player#currentPosition" -> {
                val arguments = HashMap<String, Any>()
                arguments["currentPosition"] = videoView?.currentPosition ?: 0
                result.success(arguments)
            }
            "player#isPlaying" -> {
                val arguments = HashMap<String, Any>()
                arguments["isPlaying"] = videoView?.isPlaying ?: false
                result.success(arguments)
            }
            "player#seekTo" -> {
                val position: Int? = call.argument("position")
                if (position != null)
                    videoView?.seekTo(position)
                result.success(null)
            }
            "player#toggleSound" -> {
                mute = !mute
                configureVolume()
                result.success(null)
            }
            "player#setVolume" -> {
                val volume: Double? = call.argument("volume")
                if (volume != null) {
                    this.mute = false
                    this.volume = volume
                    configureVolume()
                }
                result.success(null)
            }
        }
    }

    override fun onCreate(owner: LifecycleOwner) {
        super.onCreate(owner)
        this.configurePlayer()
    }

    override fun onPause(owner: LifecycleOwner) {
        super.onPause(owner)
        if (disposed) return
        this.pausePlayback()
    }

    override fun onStop(owner: LifecycleOwner) {
        super.onStop(owner)
        if (disposed) return
        this.stopPlayback()
    }

    override fun onDestroy(owner: LifecycleOwner) {
        super.onDestroy(owner)
        if (disposed) return
        this.destroyVideoView()
    }

    private fun configurePlayer() {
        videoView = constraintLayout.findViewById(R.id.native_video_view)
        videoView?.setOnPreparedListener(this)
        videoView?.setOnErrorListener(this)
        videoView?.setOnCompletionListener(this)
        videoView?.setZOrderOnTop(true)
        if (requestAudioFocus && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            videoView?.setAudioFocusRequest(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_EXCLUSIVE)
        }
    }

    private fun initVideo(dataSource: String?) {
        this.configurePlayer()
        if (dataSource != null) {
            val videoUri = Uri.parse(dataSource)
            this.videoView?.setVideoURI(videoUri)
            this.dataSource = dataSource
        }
    }

    private fun startPlayback() {
        if (playerState != PlayerState.PLAYING && dataSource != null) {
            if (playerState != PlayerState.NOT_INITIALIZED) {
                playerState = PlayerState.PLAYING
                videoView?.start()
            } else {
                playerState = PlayerState.PLAY_WHEN_READY
                initVideo(dataSource)
            }
        }
    }

    private fun pausePlayback() {
        val canPause = videoView?.canPause() ?: false
        if (canPause) {
            videoView?.pause()
            playerState = PlayerState.PAUSED
        }
    }

    private fun stopPlayback() {
        videoView?.stopPlayback()
        playerState = PlayerState.NOT_INITIALIZED
    }

    private fun destroyVideoView() {
        videoView?.stopPlayback()
        videoView?.setOnPreparedListener(null)
        videoView?.setOnErrorListener(null)
        videoView?.setOnCompletionListener(null)
    }

    private fun configureVolume() {
        if (mediaPlayer != null) {
            if (mute) {
                mediaPlayer?.setVolume(0f, 0f)
            } else {
                mediaPlayer?.setVolume(volume.toFloat(), volume.toFloat())
            }
        }
    }

    override fun onCompletion(mediaPlayer: MediaPlayer?) {
        this.mediaPlayer = null
        stopPlayback()
        methodChannel.invokeMethod("player#onCompletion", null)
    }

    override fun onError(mediaPlayer: MediaPlayer?, what: Int, extra: Int): Boolean {
        dataSource = null
        this.mediaPlayer = null
        playerState = PlayerState.NOT_INITIALIZED
        val arguments = HashMap<String, Any>()
        arguments["what"] = what
        arguments["extra"] = extra
        methodChannel.invokeMethod("player#onError", arguments)
        return true
    }

    override fun onPrepared(mediaPlayer: MediaPlayer?) {
        this.mediaPlayer = mediaPlayer
        configureVolume()
        if (playerState == PlayerState.PLAY_WHEN_READY) {
            this.startPlayback()
        } else {
            notifyPlayerPrepared(mediaPlayer)
        }
    }

    private fun notifyPlayerPrepared(mediaPlayer: MediaPlayer?) {
        val arguments = HashMap<String, Any>()
        if (mediaPlayer != null) {
            arguments["height"] = mediaPlayer.videoHeight
            arguments["width"] = mediaPlayer.videoWidth
            arguments["duration"] = mediaPlayer.duration
        }
        playerState = PlayerState.PREPARED
        methodChannel.invokeMethod("player#onPrepared", arguments)
    }
}

File: ios/Classes/NativeVideoViewController.swift
//
//  NativeVideoViewController.swift
//  Runner
//
//  Created by Luis Jara Castillo on 11/4/19.
//

import Foundation
import Flutter
import AVFoundation

public class NativeVideoViewController: NSObject, FlutterPlatformView {
    private var viewId: Int64
    private var methodChannel: FlutterMethodChannel
    private var videoView: VideoView?
    private var requestAudioFocus: Bool = true
    private var mute: Bool = false
    private var volume: Double = 1.0

    init(frame: CGRect, viewId: Int64, registrar: FlutterPluginRegistrar) {
        self.viewId = viewId
        videoView = VideoView(frame: frame)
        methodChannel = FlutterMethodChannel(name: "native_video_view_\(viewId)", binaryMessenger: registrar.messenger())
        super.init()
        videoView?.addOnPreparedObserver {
            [weak self] () -> Void in
            self?.onPrepared()
        }
        videoView?.addOnFailedObserver {
            [weak self] (message: String) -> Void in
            self?.onFailed(message: message)
        }
        videoView?.addOnCompletionObserver {
            [weak self] () -> Void in
            self?.onCompletion()
        }
        methodChannel.setMethodCallHandler {
            [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            self?.handle(call: call, result: result)
        }
    }

    deinit {
//        videoView?.removeFromSuperview()
        videoView = nil
        methodChannel.setMethodCallHandler(nil)
        NotificationCenter.default.removeObserver(self)
    }

    public func view() -> UIView {
        return videoView!
    }

    func handle(call: FlutterMethodCall, result: FlutterResult) -> Void {
        switch (call.method) {
        case "player#setVideoSource":
            let arguments = call.arguments as? [String: Any]
            if let args = arguments {
                self.requestAudioFocus = args["requestAudioFocus"] as? Bool ?? false
                let videoPath: String? = args["videoSource"] as? String
                if let path = videoPath {
                    let sourceType: String? = args["sourceType"] as? String
                    let isUrl = sourceType == "VideoSourceType.network"
                    configurePlayer()
                    videoView?.configure(videoPath: path, isURL: isUrl)
                }
            }
            result(nil)
            break
        case "player#start":
            videoView?.play()
            result(nil)
            break
        case "player#pause":
            videoView?.pause(restart: false)
            result(nil)
            break
        case "player#stop":
            videoView?.stop()
            result(nil)
            break
        case "player#currentPosition":
            var arguments = Dictionary<String, Any>()
            arguments["currentPosition"] = videoView?.getCurrentPosition()
            result(arguments)
            break
        case "player#isPlaying":
            var arguments = Dictionary<String, Any>()
            arguments["isPlaying"] = videoView?.isPlaying()
            result(arguments)
            break
        case "player#seekTo":
            let arguments = call.arguments as? [String: Any]
            if let args = arguments {
                let position: Int64? = args["position"] as? Int64
                videoView?.seekTo(positionInMillis: position)
            }
            result(nil)
            break
        case "player#toggleSound":
            mute = !mute
            configureVolume()
            result(nil)
            break
        case "player#setVolume":
            let arguments = call.arguments as? [String: Any]
            if let args = arguments {
                if let volume = args["volume"] as? Double {
                    mute = false
                    self.volume = volume
                    configureVolume()
                }
            }
            result(nil)
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }

    func configurePlayer() {
        handleAudioFocus()
        configureVolume()
    }

    func handleAudioFocus() {
        do {
            if requestAudioFocus {
//                try AVAudioSession.sharedInstance().setCategory(AVAudioSession.Category.ambient, mode: AVAudioSession.Mode.default, options: AVAudioSession.CategoryOptions.mixWithOthers)
                try AVAudioSession.sharedInstance().setCategory(AVAudioSession.Category.ambient)
            } else {
                try AVAudioSession.sharedInstance().setCategory(AVAudioSession.Category.playback)
            }
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print(error)
        }
    }

    func configureVolume() {
        if mute {
            videoView?.setVolume(volume: 0.0)
        } else {
            videoView?.setVolume(volume: volume)
        }
    }

    func onCompletion() {
        videoView?.stop()
        methodChannel.invokeMethod("player#onCompletion", arguments: nil)
    }

    func onPrepared() {
        var arguments = Dictionary<String, Any>()
        let height = videoView?.getVideoHeight()
        let width = videoView?.getVideoWidth()
        arguments["duration"] = videoView?.getDuration()
        arguments["height"] = height
        arguments["width"] = width
        methodChannel.invokeMethod("player#onPrepared", arguments: arguments)
    }

    func onFailed(message: String) {
        var arguments = Dictionary<String, Any>()
        arguments["message"] = message
        methodChannel.invokeMethod("player#onError", arguments: arguments)
    }
}
